### ISSUE TITLE
Jolly Jetblack Camel - LMPVaultRouter mint and deposit entry-points can be blocked by anyone

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-06-tokemak-judging#241]
Jolly Jetblack Camel

medium

# LMPVaultRouter mint and deposit entry-points can be blocked by anyone
## Summary

An attacker can easily block all mint and deposit related methods of  `LMPVaultRouter` by exploiting a public `approve(...)` method to set non-zero token allowances on behalf of the router.   

As a consequence, most **main user entry-points** of the protocol can be **arbitrarily blocked** by **anyone** via a simple and unrestrained attack path.

## Vulnerability Detail

The `SafeERC20.safeApprove(...)` method, which does not allow approval from from *non-zero* to *non-zero* allowance,
```solidity
    function safeApprove(
        IERC20 token,
        address spender,
        uint256 value
    ) internal {
        // safeApprove should only be called when setting an initial allowance,
        // or when resetting it to zero. To increase and decrease it, use
        // 'safeIncreaseAllowance' and 'safeDecreaseAllowance'
        require(
            (value == 0) || (token.allowance(address(this), spender) == 0),
            "SafeERC20: approve from non-zero to non-zero allowance"
        );
        _callOptionalReturn(token, abi.encodeWithSelector(token.approve.selector, spender, value));
    }
```
is used by [LMPVaultRouterBase.mint(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouterBase.sol#L22-L41) in [L35](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouterBase.sol#L35) and [LMPVaultRouterBase_.deposit(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouterBase.sol#L59-L70) in [L66](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouterBase.sol#L66) via the base method [PeripheryPayments.approve(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/utils/PeripheryPayments.sol#L35-L37).  

However, [PeripheryPayments.approve(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/utils/PeripheryPayments.sol#L35-L37) is a *public* method which is inherited by `LMPVaultRouterBase` and subsequently `LMPVaultRouter`.  
As a consquence, **anyone** can give *non-zero* approval, on behalf of the `LMPVaultRouter` contract, to **any** vault for **any** ERC-20 token.  
Therefore, it's fairly easy for an attacker to make the `SafeERC20.safeApprove(...)` calls within [LMPVaultRouterBase.mint(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouterBase.sol#L22-L41) and [LMPVaultRouterBase_.deposit(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouterBase.sol#L59-L70) revert with `SafeERC20: approve from non-zero to non-zero allowance`.

## Impact

The following top-level methods are subject to DoS and loss of gas fee (due to failed transaction) because of the above vulnerability:
1. [LMPVaultRouter(Base).mint(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouterBase.sol#L22-L41)
2. [LMPVaultRouter(Base).deposit(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouterBase.sol#L43-L57) which calls `LMPVaultRouterBase_.deposit(...)`
3. [LMPVaultRouter.withdrawToDeposit(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouter.sol#L26-L37) which calls `LMPVaultRouterBase_.deposit(...)`
4. [LMPVaultRouter.swapAndDepositToVault(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouter.sol#L39-L60) which calls `LMPVaultRouterBase_.deposit(...)`
5. [LMPVaultRouter.redeemToDeposit(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouter.sol#L62-L73) which calls `LMPVaultRouterBase_.deposit(...)`
6. [LMPVaultRouter.depositMax(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/vault/LMPVaultRouter.sol#L75-L87) which calls `LMPVaultRouterBase_.deposit(...)`

Consequently, every user, or even worse every contract (less flexibility), that relies on one of the above methods is subject to this DoS attack.  
Due to the fact that this vulnerability is easy and cheap to exploit (set allowances for all vaults within one transaction) and an attacker also being able to approve-front-run transactions at will, the resulting DoS scenario is *severe* and can be considered *permanent*.  

Furthermore, the affected contract and its methods are the main entry point of the protocol, see [README](https://github.com/Tokemak/v2-core-audit-2023-07-14/tree/62445b8ee3365611534c96aef189642b721693bf):
```text
End-Users View

A user will largely only interact with the Router directly (the exception being rewards+staking).
The Router includes safety and convenience methods for interacting with the ERC4626 compatible Autopilot Vaults.
From a safety perspective, this includes slippage-based variants of deposit/mint/redeem/withdraw(), and for migrating between vaults, redeemToDeposit().

The only time a user should be interacting with the vault directly is when claiming rewards.
```

## Code Snippet

The following PoC modifies the two existing test cases
* `LMPVaultRouterTest.test_deposit()` and
* `LMPVaultRouterTest.test_mint()`
in order to demonstrate the above vulnerability.  

Just apply the *diff* below and run the tests with `forge test -vv --match-contract LMPVaultRouterTest`:
```diff
diff --git a/test/vault/LMPVaultRouter.t.sol b/test/vault/LMPVaultRouter.t.sol
index 93809f8..7b3f7c3 100644
--- a/test/vault/LMPVaultRouter.t.sol
+++ b/test/vault/LMPVaultRouter.t.sol
@@ -126,7 +126,10 @@ contract LMPVaultRouterTest is BaseTest {
         vm.expectRevert(abi.encodeWithSelector(ILMPVaultRouterBase.MinSharesError.selector));
         lmpVaultRouter.deposit(lmpVault, address(this), amount, minSharesExpected);
 
-        // -- now do a successful scenario -- //
+        // -- now do a successful scenario ... if there wasn't a griefer -- //
+        vm.prank(makeAddr("griefer"));
+        lmpVaultRouter.approve(baseAsset, address(lmpVault), 1);
+
         _deposit(lmpVault, amount);
     }
 
@@ -159,7 +162,10 @@ contract LMPVaultRouterTest is BaseTest {
         vm.expectRevert(abi.encodeWithSelector(ILMPVaultRouterBase.MaxAmountError.selector));
         lmpVaultRouter.mint(lmpVault, address(this), amount, maxAssets);
 
-        // -- now do a successful mint scenario -- //
+        // -- now do a successful mint scenario ... if there wasn't a griefer -- //
+        vm.prank(makeAddr("griefer"));
+        lmpVaultRouter.approve(baseAsset, address(lmpVault), 1);
+
         _mint(lmpVault, amount);
     }
 
```

## Tool used

Manual Review

## Recommendation

Make the [PeripheryPayments.approve(...)](https://github.com/Tokemak/v2-core-audit-2023-07-14/blob/62445b8ee3365611534c96aef189642b721693bf/src/utils/PeripheryPayments.sol#L35-L37)  method *internal*, since it's not used elsewhere:

```diff
diff --git a/src/utils/PeripheryPayments.sol b/src/utils/PeripheryPayments.sol
index 24dcde0..2fbd296 100644
--- a/src/utils/PeripheryPayments.sol
+++ b/src/utils/PeripheryPayments.sol
@@ -32,7 +32,7 @@ abstract contract PeripheryPayments {
 
     receive() external payable { }
 
-    function approve(IERC20 token, address to, uint256 amount) public payable {
+    function approve(IERC20 token, address to, uint256 amount) internal {
         token.safeApprove(to, amount);
     }
 

```

**Comment by nevillehuang** [comment]
Escalate

> As a consquence, anyone can give non-zero approval, on behalf of the LMPVaultRouter contract, to any vault for any ERC-20 token.

This is a clear duplicate of #346

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> > As a consquence, anyone can give non-zero approval, on behalf of the LMPVaultRouter contract, to any vault for any ERC-20 token.
> 
> This is a clear duplicate of #346 

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Trumpero** [comment]
Agree this issue is a duplicate of #346

**Comment by Evert0x** [comment]
Planning to accept escalation and mark as a duplicate

**Comment by Evert0x** [comment]
Result:
Medium
Duplicate of #346

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [nevillehuang](https://github.com/sherlock-audit/2023-06-tokemak-judging/issues/241/#issuecomment-1745322858): accepted
